addons/product/models/product_product.py

```diff
@@ -533,50 +533,73 @@ def get_display_name(name, code):
-            if not product_ids and operator not in expression.NEGATIVE_TERM_OPERATORS:
-                # Do not merge the 2 next lines into one single search, SQL search performance would be abysmal
-                # on a database with thousands of matching products, due to the huge merge+unique needed for the
-                # OR operator (and given the fact that the 'name' lookup results come from the ir.translation table
-                # Performing a quick memory merge of ids in Python will give much better performance
-                product_ids = list(self._search(domain + [('default_code', operator, name)], limit=limit, order=order))
-                if not limit or len(product_ids) < limit:
-                    # we may underrun the limit because of dupes in the results, that's fine
-                    limit2 = (limit - len(product_ids)) if limit else False
-                    product2_ids = self._search(domain + [('name', operator, name), ('id', 'not in', product_ids)], limit=limit2, order=order)
-                    product_ids.extend(product2_ids)
-            elif not product_ids and operator in expression.NEGATIVE_TERM_OPERATORS:
-                domain2 = expression.OR([
-                    ['&', ('default_code', operator, name), ('name', operator, name)],
-                    ['&', ('default_code', '=', False), ('name', operator, name)],
-                ])
-                domain2 = expression.AND([domain, domain2])
-                product_ids = list(self._search(domain2, limit=limit, order=order))
-            if not product_ids and operator in positive_operators:
-                ptrn = re.compile(r'(\[(.*?)\])')
-                res = ptrn.search(name)
-                if res:
-                    product_ids = list(self._search([('default_code', '=', res.group(2))] + domain, limit=limit, order=order))
-            # still no results, partner in context: search on supplier info as last hope to find something
-            if not product_ids and self._context.get('partner_id'):
-                suppliers_ids = self.env['product.supplierinfo']._search([
-                    ('partner_id', '=', self._context.get('partner_id')),
-                    '|',
-                    ('product_code', operator, name),
-                    ('product_name', operator, name)])
-                if suppliers_ids:
-                    product_ids = self._search([('product_tmpl_id.seller_ids', 'in', suppliers_ids)], limit=limit, order=order)
+    def _search_display_name(self, operator, value):
+        if operator in expression.NEGATIVE_TERM_OPERATORS:
+            positive_operator = expression.TERM_OPERATORS_NEGATION[operator]
         else:
-            product_ids = self._search(domain, limit=limit, order=order)
-        return product_ids
+            positive_operator = operator
+        combine = expression.OR if positive_operator == operator else expression.AND
```

Comment:
The use of `positive_operator` in method `_search_display_name` only makes code harder to read.
All you need is a boolean for the condition `operator in expression.NEGATIVE_TERM_OPERATORS`.